### PR TITLE
Option to disallow explicit Any types (--disallow-any=explicit)

### PR DESCRIPTION
**Pull request by ilinum**
Type aliases are a little tricky to deal with because they get "inlined".
The solution was to unmark Anys as explicit right after we check a type alias
for explicit Anys.

### CODE REVIEW

File: mypy/messages.py [hunk 1]
```diff
@@ -872,6 +872,9 @@ def unimported_type_becomes_any(self, prefix: str, typ: Type, ctx: Context) -> N
         self.fail("{} becomes {} due to an unfollowed import".format(prefix, self.format(typ)),
                   ctx)
 
+    def explicit_any(self, location: str, ctx: Context) -> None:
+        self.fail('Explicit "Any" is not allowed in a {}'.format(location), ctx)
```
Comment (ddfisher) [inline mypy/messages.py]: I wonder if this error would be clearer if it was just `Explicit "Any" is not allowed.`.  This reads to me like it's implying that the `Any` is not allowed in that particular location, but in fact it's not allowed in _any_ location.

File: mypy/semanal.py [hunk 2]
```diff
@@ -4310,3 +4334,21 @@ def find_fixed_callable_return(expr: Expression) -> Optional[CallableType]:
             if isinstance(t.ret_type, CallableType):
                 return t.ret_type
     return None
+
+
+def unmark_any_as_explicit(t: Type) -> None:
```
Comment (ddfisher) [inline mypy/semanal.py]: I think this name (and that of the associated class) needs a bit more thought -- I don't think it's clear what it means outside of the context of this PR.

File: mypy/types.py [hunk 3]
```diff
@@ -254,13 +254,16 @@ class AnyType(Type):
...
+                 explicit: bool = False,
                  line: int = -1,
                  column: int = -1) -> None:
         super().__init__(line, column)
         # Was this Any type was inferred without a type annotation?
         self.implicit = implicit
         # Does this come from an unfollowed import? See --disallow-any=unimported option
         self.from_unimported_type = from_unimported_type
+        # Does this Any come from an explicit type annotation?
+        self.explicit = explicit
```
Comment (ddfisher) [inline mypy/types.py]: Is this not always the opposite of `implicit`?  If not, the specifics need to be made very clear in the comment.

File: mypy/semanal.py [hunk 4]
```diff
@@ -4310,3 +4334,21 @@ def find_fixed_callable_return(expr: Expression) -> Optional[CallableType]:
...
+    """For all types within t, if that type is Any, set explicit to False"""
+    t.accept(UnmarkAnyAsExplicit())
+
+
+class UnmarkAnyAsExplicit(TypeQuery[None]):
```
Comment (ddfisher) [inline mypy/semanal.py]: This is a bit of an abuse of the `TypeQuery` class.  None of its parent classes seem to be set up for doing this easily, but I'm not entirely sure how I feel about this overall.  @JukkaL do you think this is a reasonable thing to do?

File: test-data/unit/cmdline.test [hunk 5]
```diff
@@ -681,3 +681,178 @@ from unreal import F
...
+
+[file mypy.ini]
+[[mypy]
+[[mypy-m*]
+disallow_any = explicit
+
+[file m.py]
+from typing import Any, List
+v: Any = ''
+w = ['']  # type: List[Any]
```
Comment (ddfisher) [inline test-data/unit/cmdline.test]: Nit: It's better to vary test cases along one axis at once.  Here, you're switching from testing a type annotation to a type comment, but you're also switching from testing a plain `Any` to an `Any` in a generic class.  (And also the list shouldn't need any contents.)

File: test-data/unit/cmdline.test [hunk 6]
```diff
@@ -681,3 +681,178 @@ from unreal import F
...
+[file m.py]
+from typing import Any, List, TypeVar, Tuple
+
+T = TypeVar('T')
+
+TupleAny = Tuple[Any, T]  # error
+
+def foo(x: TupleAny[str]) -> None:  # no error
+    pass
+
```
Comment (ddfisher) [inline test-data/unit/cmdline.test]: Could you also check that `TupleAny[Any]` produces an error?

File: test-data/unit/cmdline.test [hunk 7]
```diff
@@ -681,3 +681,178 @@ from unreal import F
...
+[file mypy.ini]
+[[mypy]
+[[mypy-m*]
+disallow_any = explicit
+
+[file m.py]
+from typing import Any, List, NewType
+
+# The commented out line is covered without the flag enabled - it is tested elsewhere
```
Comment (ddfisher) [inline test-data/unit/cmdline.test]: I'd instead be a bit more specific and say "making a NewType of Any is never allowed".

File: mypy/types.py (hunk 3 above)
Comment (ilinum) [inline mypy/types.py]: Yes, `explicit` is not always the opposite of `implicit`, which is confusing.

`implicit` refers to situations when `Any` was inferred because a variable or parameter was not annotated.
`explicit` refers to situations when there was an explicit `Any` annotation. 

However, if this `Any` comes, for instance, from an unimported type, both `explicit` and `implicit` will be `False`.

It seems to me that `implicit` has a wrong name. It is currently used for `--disallow-untyped-defs` to see if this `Any` comes from a parameter or variable with no explicit type annotation. 

I think instead of tweaking comments we should rename `implicit`. How do you feel about calling it `from_unannotated`?

File: mypy/messages.py (hunk 1 above)
Comment (ilinum) [inline mypy/messages.py]: Yeah, actually, I agree. I think I was trying to be too fancy when it wasn't necessary.

File: mypy/semanal.py (hunk 4 above)
Comment (ilinum) [inline mypy/semanal.py]: Yeah, I agree, this does not seem right.
Moreover, this mutates a type, which is also not great.

What I need is a TypeVisitor that visit all types within a type. Maybe I can refactor one from `TypeQuery` and then make `TypeQuery` inherit the new visitor or something?

File: mypy/semanal.py (hunk 2 above)
Comment (ilinum) [inline mypy/semanal.py]: Yeah, naming things is hard - I went through several iterations of names trying to name this function. 

How about `unset_any_type_as_explicit`? Alternatively, we could try to make the function more generic and add a `bool` parameter `explicit`. Something like:
```py
def set_any_type_explicit_attribute(explicit: bool) -> None: 
    ...
```

File: test-data/unit/cmdline.test (hunk 5 above)
Comment (ilinum) [inline test-data/unit/cmdline.test]: All right, that makes a lot of sense.

Also, I will split these tests into two.

File: test-data/unit/cmdline.test (hunk 7 above)
Comment (ilinum) [inline test-data/unit/cmdline.test]: Yep, that wording is definitely better. 

However, I think it's better to include the test (`Baz = NewType('Baz', Any)`).
In a different PR Guido [pointed out](https://github.com/python/mypy/pull/3580/files/0cd8e4a67155c2cd526f325ad18f86b4b7bf6a52#r123151104) that tests should follow features, not implementation.

And I think in this case, it's better to have an extra tests. Because if this errors stops being reported `--disallow-any=explicit` will not work - it won't report an explicit Any. 

So I am going to include the test here.

File: test-data/unit/cmdline.test (hunk 6 above)
Comment (ilinum) [inline test-data/unit/cmdline.test]: Wow, this is an interesting edge case. Will add a test.

**Comment by ilinum** [comment]
Most feedback should be addressed now. 

I also realized that the originally this PR did not report errors when `Any` was used in `TypedDict` declaration, so I [added](https://github.com/python/mypy/pull/3579/commits/e3cd6ebf17bab68caf28998d3011e686cea48f03) that.

File: mypy/types.py (hunk 3 above)
Comment (ilinum) [inline mypy/types.py]: Added comment on implicit to make it more clear.

**Comment by ilinum** [comment]
All feedback should be addressed now, except for using `TypeQuery` to set `AnyType` as non-explicit.

File: mypy/semanal.py [hunk 8]
```diff
@@ -4381,3 +4410,21 @@ def find_fixed_callable_return(expr: Expression) -> Optional[CallableType]:
...
                 return t.ret_type
     return None
+
+
+def mark_any_non_explicit(t: Type) -> None:
+    """For all types within t, if that type is Any, set explicit to False"""
+    t.accept(MarkAnyNonExplicit())
+
+
+class MarkAnyNonExplicit(TypeQuery[None]):
```
Comment (JukkaL) [inline mypy/semanal.py]: What about having a `TypeQuery` that returns all `Any` types that are part of a type, and then iterate over the results of the query and set `explicit`? I.e. something like this:

```py
for any_type in find_all_any_types(t):
    any_type.explicit = False
```

Comment (JukkaL) [inline mypy/semanal.py]: Hmm actually that might not be a good idea, since mutating types is only safe if you are certain that there is ever only a single reference to the type object. What about a type transform that replaces `Any` instances with a copy that has `explicit` set to `False`? You should be able to use `TypeTranslator`.

File: mypy/checker.py [hunk 9]
```diff
@@ -627,6 +627,11 @@ def is_implicit_any(t: Type) -> bool:
                                 if has_any_from_unimported_type(arg_type):
                                     prefix = "Argument {} to \"{}\"".format(idx + 1, fdef.name())
                                     self.msg.unimported_type_becomes_any(prefix, arg_type, fdef)
+                    if ('explicit' in self.options.disallow_any and
```
Comment (JukkaL) [inline mypy/checker.py]: This same piece of code gets repeated almost identically many times. Can you refactor this into a function or a few functions?

Comment (ilinum) [inline mypy/checker.py]: Yep, done!

File: mypy/semanal.py (hunk 8 above)
Comment (ilinum) [inline mypy/semanal.py]: Yes, that's a really good idea!

Done!

**Comment by ilinum** [comment]
All feedback is addressed now, this PR should be ready to be merged.
